src/vmm/src/vstate/vcpu/x86_64.rs

```diff
@@ -0,0 +1,460 @@
+            xsave,
+        })
+    }
+
+    /// Use provided state to populate KVM internal state.
+    pub fn restore_state(&self, state: &VcpuState) -> Result<()> {
+        /*
+         * Ordering requirements:
+         *
+         * KVM_GET_VCPU_EVENTS/KVM_SET_VCPU_EVENTS is unsafe if other vCPUs are
+         * still running.
+         *
+         * Some SET ioctls (like set_mp_state) depend on kvm_vcpu_is_bsp(), so
+         * if we ever change the BSP, we have to do that before restoring anything.
+         * The same seems to be true for CPUID stuff.
+         *
+         * SREGS saves/restores a pending interrupt, similar to what
+         * VCPU_EVENTS also does.
+         *
+         * SET_REGS clears pending exceptions unconditionally, thus, it must be
+         * done before SET_VCPU_EVENTS, which restores it.
+         *
+         * SET_LAPIC must come after SET_SREGS, because the latter restores
+         * the apic base msr.
+         *
+         * SET_LAPIC must come before SET_MSRS, because the TSC deadline MSR
+         * only restores successfully, when the LAPIC is correctly configured.
+         */
+        self.fd
+            .set_cpuid2(&state.cpuid)
+            .map_err(Error::VcpuSetCpuid)?;
+        self.fd
+            .set_mp_state(state.mp_state)
+            .map_err(Error::VcpuSetMpState)?;
+        self.fd.set_regs(&state.regs).map_err(Error::VcpuSetRegs)?;
+        self.fd
+            .set_sregs(&state.sregs)
+            .map_err(Error::VcpuSetSregs)?;
+        self.fd
+            .set_xsave(&state.xsave)
+            .map_err(Error::VcpuSetXsave)?;
+        self.fd.set_xcrs(&state.xcrs).map_err(Error::VcpuSetXcrs)?;
+        self.fd
+            .set_debug_regs(&state.debug_regs)
+            .map_err(Error::VcpuSetDebugRegs)?;
+        self.fd
+            .set_lapic(&state.lapic)
+            .map_err(Error::VcpuSetLapic)?;
+        self.fd.set_msrs(&state.msrs).map_err(Error::VcpuSetMsrs)?;
+        self.fd
+            .set_vcpu_events(&state.vcpu_events)
+            .map_err(Error::VcpuSetVcpuEvents)?;
+        Ok(())
+    }
+
+    /// Runs the vCPU in KVM context and handles the kvm exit reason.
+    ///
+    /// Returns error or enum specifying whether emulation was handled or interrupted.
+    pub fn run_emulation(&self, exit: VcpuExit) -> super::Result<VcpuEmulation> {
+        match exit {
+            VcpuExit::IoIn(addr, data) => {
+                if let Some(pio_bus) = &self.pio_bus {
+                    pio_bus.read(u64::from(addr), data);
+                    METRICS.vcpu.exit_io_in.inc();
+                }
+                Ok(VcpuEmulation::Handled)
+            }
+            VcpuExit::IoOut(addr, data) => {
+                if let Some(pio_bus) = &self.pio_bus {
+                    pio_bus.write(u64::from(addr), data);
+                    METRICS.vcpu.exit_io_out.inc();
+                }
+                Ok(VcpuEmulation::Handled)
+            }
+            r => {
```

Comment:
nit: I would rename `r` to something like `unexpected_exit`.

Reply:
Renamed.